apex/normalization/instance_norm.py

```diff
@@ -0,0 +1,808 @@
+        [batch_dim, channel_dim],
+    )
+
+    invstd_bcast = fd.ops.broadcast_in_dim(
+        invstd,
+        extent,
+        [batch_dim, channel_dim],
+    )
+    grad_scale = (
+        invstd_bcast
+        if weight is None
+        else fd.ops.mul(
+            invstd_bcast,
+            fd.ops.broadcast_in_dim(weight, extent, [0]),
+        )
+    )
+    if use_input_stats:
+        proj = fd.ops.mul(fd.ops.sub(x, mean), proj_scale)
+        grad_input = fd.ops.mul(
+            fd.ops.sub(
+                fd.ops.sub(grad_output, proj),
+                grad_mean,
+            ),
+            grad_scale,
+        )
+    else:
+        grad_input = fd.ops.mul(grad_output, grad_scale)
+
+    if weight is not None:
+        grad_weight = fd.ops.mul(dot_p, invstd)
+        grad_weight_reduced = fd.ops.sum(grad_weight, [0])
+    else:
+        grad_weight_reduced = None
+    if bias is not None:
+        grad_bias = grad_output_sum
+        grad_bias_reduced = fd.ops.sum(grad_bias, [0])
+    else:
+        grad_bias_reduced = None
+
+    return grad_input, grad_weight_reduced, grad_bias_reduced
+
+
+# The fusion caches for forward and backward evaluation map "keys" in the form
+# of tuples of parameters, to Fusion objects. These Fusion objects hold
+# FusionExecutorCache objects under the hood, which are able to repeatedly
+# execute a Fusion without redefining or rescheduling them.
+_forward_fusion_cache: Dict[
+    Tuple[
+        Tuple[NamedAxis, ...],
+        DataType,
+        DataType,
+        DataType,
+        DataType,
+        int,
+        bool,
+        bool,
+        bool,
+        bool,
+    ],
+    Fusion,
+] = {}
+_backward_fusion_cache: Dict[
+    Tuple[
+        Tuple[NamedAxis, ...],
+        DataType,
+        DataType,
+        DataType,
+        DataType,
+        int,
+        bool,
+        bool,
+        bool,
+        bool,
+    ],
+    Fusion,
+] = {}
+
+
+class NormNVFuserFunction(torch.autograd.Function):  # type: ignore
+    @staticmethod
+    def forward(
+        ctx: Any,  # contexts are actually objects of the type we are currently defining
+        x: torch.Tensor,
+        weight: Optional[torch.Tensor],
+        bias: Optional[torch.Tensor],
+        running_mean: Optional[torch.Tensor],
+        running_var: Optional[torch.Tensor],
+        use_input_stats: bool,
+        momentum: float,
+        eps: float,
+        unbiased: bool,
+        stat_axes: List[NamedAxis],
+    ) -> torch.Tensor:
+        channels_last = x.is_contiguous(
+            memory_format=torch.channels_last
+        ) or x.is_contiguous(memory_format=torch.channels_last_3d)
+        xorig = x
+        if channels_last:
+            order = [0] + [i for i in range(2, len(x.shape))] + [1]
+            x = x.permute(order)
+        assert x.is_contiguous()
+
+        x_datatype = torch2datatype(x.dtype)
+
+        # The cache key for this Fusion definition. Misses will trigger redefinition and rescheduling.
+        key = (
```

Comment:
if the fusion definition above is shape specific, then input shape here should be part of the cache entry.

Reply:
You're right. I've removed the cache anyway now though (see above).

Reply:
After talking with Kevin, I'm removing the explicit cache altogether. It was based on `Fusion`s which are no longer exposed, and was probably misconceived by me to begin with since the `FusionDefinition` has its own cache mechanism, so the explicit cache here only saves the actual building of the FusionDefinition (stuff in the `with` block). `FusionDefinition` is lightweight so that part is fast anyway.